File: c/eu018.c
// project euler (projecteuler.net) problem 18
// solution by Kevin Retzke (retzkek@gmail.com)
// November 2012

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#include "pe.h"

typedef struct {
	int **t;
	int size;
} triangle;

// read a triangle from a text file.
// Example (size=4):
//
//	3
//	7 5
//	2 4 6
//	8 5 9 3
triangle *readTriangleFromFile(char *filename, int size) {
	FILE *f;
	triangle *t;

	t = malloc(sizeof(triangle));
	t->size = size;
	t->t = malloc(size*sizeof(int*));

	f = fopen(filename, "r");
	for (int j = 0; j < size; j++) {
		t->t[j] = malloc((j+1)*sizeof(int));
		for (int i = 0; i < j+1; i++) {
			fscanf(f, "%d", &t->t[j][i]);
		}
	}
	fclose(f);
	return t;
}

// clean up triangle memory.
void cleanTriangle(triangle *t) {
	for (int j=0; j < t->size; j++) {
		free(t->t[j]);
	}
	free(t->t);
	free(t);
}

// Print prints the triangle to stdout.
void printTriangle(triangle *t) {
	for (int j = 0; j < t->size; j++) {
		for (int i = 0; i < j+1; i++) {
			printf("%3d ",t->t[j][i]);
		}
		printf("\n");
	}
}

// compute the maximum sum of all paths from top to bottom.
int maxPathSum(triangle *t) {
	int *row0, *row1;
	row0 = malloc(t->size*sizeof(int));
	for (int i=0; i<t->size; i++) {
		row0[i] = t->t[t->size-1][i];
	}
	for (int j = t->size-2; j >= 0; j--) {
		row1 = malloc((j+1)*sizeof(int));
		for (int i = 0; i < j+1; i++) {
			if (row0[i] > row0[i+1]) {
				row1[i] = t->t[j][i] + row0[i];
			} else {
				row1[i] = t->t[j][i] + row0[i+1];
			}
		}
		free(row0);
		row0 = row1;
	}
	int r = row0[0];
	free(row0);
	return r;
}

int main()
{
	triangle *t;

	t = readTriangleFromFile("../data/eu018test.dat",4);
	pe_test_eq(maxPathSum(t), 23, "%d");
	cleanTriangle(t);

	t = readTriangleFromFile("../data/eu018.dat",15);
	//printTriangle(t);
	printf("%d\n",maxPathSum(t));
	cleanTriangle(t);
}

File: python/eu018.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problems 18 and 67:
solution by Kevin Retzke (retzkek@gmail.com), April 2009

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

   3
  7 5
 2 4 6
8 5 9 3

That is, 3 + 7 + 4 + 9 = 23.

(test case saved to eu018test.dat)

PROBLEM 18:

Find the maximum total from top to bottom of the triangle below:

              75
             95 64
            17 47 82
           18 35 87 10
          20 04 82 47 65
         19 01 23 75 03 34
        88 02 77 73 07 63 67
       99 65 04 28 06 16 70 92
      41 41 26 56 83 40 80 70 33
     41 48 72 33 47 32 37 16 94 29
    53 71 44 65 25 43 91 52 97 51 14
   70 11 33 28 77 73 17 78 39 68 17 57
  91 71 52 38 17 14 91 43 58 50 27 29 48
 63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23

(data saved to eu018.dat)

PROBLEM 67:
"""

import sys

def main(argv=None):
    if argv is None:
        argv = sys.argv
    if len(argv) != 2:
       print "usage: %s [data file]" % argv[0] 
       return 1
    f = open(argv[1],'r')
    triangle = []
    for line in f:
        triangle.append(map(int, line.split()))
    size = len(triangle)
    # collapse the triangle down by picking the best choice in the last row
    # for each element in the penultimate row and sum.  rinse and repeat.
    # example:
    #       03
    #     07  05
    #   02  04  06
    # 08  05  09  03
    #
    #     03
    #   07  05
    # 10  13  15
    #
    #   03
    # 20  20
    #
    # 23
    for row in range(size-2, -1, -1):
        for col in range(row+1):
            triangle[row][col] += max(triangle[row+1][col], triangle[row+1][col+1])
    print triangle[0][0]
            


if __name__ == "__main__":
    sys.exit(main())

